### DISCUSSION TITLE
Is there any way to use SQL files in my $queryRaw calls / auto formatting SQL?

### DISCUSSION THREAD

**Discussion by leo-petrucci** [prisma/prisma#12438] [Q&A] [answered]
My current project uses a lot of Prisma's standard ORM functionality, but there are some queries that need to written in pure SQL and they're **huge**.

At the moment I've got them written in separate functions so I can reuse them:

```
// an example, my query would be 200 more lines 
export const getSomething = () => {
  return prisma.$queryRaw(sql`SELECT * FROM "Something"`);
}
```

The issue with this is that I cannot find a way to auto-format my queries. Every single VsCode/NPM package I've found only handles auto-formatting SQL that's been placed in `.sql` files.

So my questions are:

1. Can I have a separate `.sql` file and somehow import it to be used in my `prisma.$queryRaw` function?
2. Is there a way to get VSCode to auto-format queries placed inside the `sql` template string?

**Comment by Jolg42** [comment, marked as answer]
1. I think you could do this like this
```
import { readFile } from 'fs/promises';

//  'SELECT * FROM users WHERE email = $1',
const sqlFromFile =  await readFile(fileName, { encoding: 'utf8' });

prisma.$queryRawUnsafe(
  sqlFromFile,
  'helloooo@prisma.io'
)
```
Note from the docs https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#queryrawunsafe
> By using this method with user inputs (i.e. SELECT * FROM table WHERE columnx = ${userInput}), you open up the possibility for SQL injection attacks. SQL injection attacks can expose your data, be it confidential or otherwise sensitive, to being modified, or even destroyed.

2. I know that Prettier formats fenced code blocks in Markdown for example like

````
```ts
const       something      =   '   '  ;
```
````
Will turn into

````
```ts
const something = '   '
```
````

And this looks interesting/related
> The embed function is called when the plugin needs to print one language inside another. Examples of this are printing CSS-in-JS or fenced code blocks in Markdown.
https://github.com/prettier/prettier/blob/main/docs/plugins.md#optional-embed

The community prettier plugin is here https://github.com/umidbekk/prettier-plugin-prisma

It's not that clear how easy it would be to do, though 🤔

**Reply by leo-petrucci** [reply]
First method was what I was going to do, but my API is being deployed serverless so I won't have access to the FS in production, so I think that's a no-go.

I'll give the `embed` function a go, thank you 😁

**Reply by Jolg42** [reply]
Indeed. Very curious of what you find out along the way 🙌🏼 
Note that umidbekk is very friendly and could be interested / happy to help if you need help and open an issue in the repo there.

**Comment by Sec-ant** [comment]
Maybe a bit late to the party, but you can format embedded SQL languages in JS/TS files now with the plugin [`prettier-plugin-embed`](https://github.com/Sec-ant/prettier-plugin-embed).

**Reply by Jolg42** [reply]
Thanks for sharing this!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
